crypto/fipsmodule/rand/rand.c

```diff
@@ -404,7 +451,8 @@ void RAND_bytes_with_additional_data(uint8_t *out, size_t out_len,
   }
 
   if (state->calls >= kReseedInterval ||
-      state->fork_generation != fork_generation) {
+      state->fork_generation != fork_generation ||
+      state->snapsafe_generation != snapsafe_generation) {
```

Comment:
It is a good point that a snapsafe-type uniqueness breaking event (ube) is different from a fork-type ube. I think POSIX enforces your assumption: only the forking thread memory is propagated to the new process.

The time window in switch the snapsafe-type ube should occur is very small: depending on the platform (etc.) we are looking at 1-2 digit microsecond. There is also the time window after this call, where snapsafe can be violated. This time window is arguably much bigger in size.

However, since we can do better, we should do better. Since the likelihood of the snapsafe-type ube triggering after returning from randomness generation is much higher (if we consider a uniform distribution over the entire time window), I think it makes sense to have **two** checks: one before and one after. The a priori check will catch any snapsafe-type ube's between the randomness call returning and being called again, while the posteriori check will catch an unlucky snapsafe-type ube during randomness generation.

The a priori check will catch most snapsafe-type ube's, and therefore save, in majority of cases, a double randomness generation.

This check is negligible in cost. Hence, no performance issues having both.